### PR TITLE
Upgrade to VEP 111.0

### LINKED ISSUE #94: Update VEP version from 103 to 111.0
Currently we are running version 103

The latest version is 110.1

[Link to VEP repo](https://github.com/Ensembl/ensembl-vep)

### PR DESCRIPTION
**Pull request by alkc**
# TODO:

- [x] Build and setup new container
- [x] Move container pathspec out of main.nf
- [x] Move vep params out of main.nf
- [x] Clean up vep config
- [x] Install new fasta
- [x] Install new cache
- [x] Install MaxEntScan
- [x] Install CADD
- [x] LofTool
- [x] dbSNFP
- [x] Run finishes w/o errors
- [ ] #150 
- [x] Fix scout unable to parse variant frequencies
- [x] Add support for new consequences in scout: https://github.com/Clinical-Genomics/scout/pull/4434
- [ ] Might be required to update the rank models w/ missing consequences for loading into scout. If not then save for a final PR to the validation branch. If yes:
   - [ ] WGS trio
   - [ ] WGS single 
   - [ ] WGS SV trio
   - [ ] WGS SV single 
   - [ ] (strained breathing)
- [ ] Complete major change test for WGS singles and trios to ensure nothing crashes 

# Description

This update updates VEP from v103 to v111.

The update also moves the vep container pathspec out of `main.nf` to the config file.

closes #94 

## Type of change

- Major change

# Checklist:

- [x] My code follows the [style guidelines for NF pipelines at CMD](http://mtlucmds1.lund.skane.se/wiki/doku.php?id=nextflow&s[]=nextflow#code_style_at_cmd)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings (Keep an eye on `.nextflow.log` !)
- [x] I have updated the CHANGELOG
- [ ] The latest commit in the master branch is tagged
      with the correct version number

## Major change
- [x] Stub run completes without errors or new warnings
- [x] `onco` run finishes without any new warnings/errors and the results can 
       be loaded into scout
- [x] `oncov2` run finishes without any new warnings/errors and the results can 
       be loaded into scout
- [x] `wgs` single run finishes without any new warnings/errors and the results 
       can be loaded into scout
- [x] `wgs` trio run finishes without any new warnings/errors and the results 
       can be loaded into scout
- [x] At least one other person has reviewed and approved my code
- [x] I have made corresponding changes to the documentation

# Instructions for the reviewers

## [Optional] Additional information

# Review

## Review performed by:
- Reviewer 1: Jakob @Jakob37 
    
## Testing performed by:
- Alex @alkc 
- Jakob @Jakob37 

# Post-merge
 - [ ] #142 
 - [ ] #150

### CODE REVIEW

**Comment by alkc** [comment]
vep 111 fasta seems to differ from vep 98 fasta that we use. switching it out.

**Comment by alkc** [comment]
at this stage the pipeline finishes a wgs run w/o new warnings and errors. ready for testing/validation.

**Comment by Jakob37** [comment]
Btw, I guess we need to keep the versions updated in the list of softwares:

https://github.com/Clinical-Genomics-Lund/nextflow_wgs/blob/master/newdocs/list_of_all_used_software.md

**Comment by alkc** [comment]
> Btw, I guess we need to keep the versions updated in the list of softwares:
> 
> https://github.com/Clinical-Genomics-Lund/nextflow_wgs/blob/master/newdocs/list_of_all_used_software.md

Ah, another item for my glorious checklists!

**Comment by alkc** [comment]
VEP 105 changelog reports:

3 new Sequence Ontology terms are reported for more detailed splice consequence annotation:
- splice_donor_5th_base_variant (SO:0001787)
- splice_donor_region_variant (SO:0002170)
- splice_polypyrimidine_tract_variant (SO:0002169)

These three guys need to be included in the rank models. Should find out if they completely replace `splice_donor_variant` annotation.

**Comment by alkc** [comment]
The new annotations probably need to be ranked in `modify_scout_vcf.pl`

https://github.com/Clinical-Genomics-Lund/nextflow_wgs/blob/b9eff369ae87bbd5f38829b9f8b57cf969ea24b5/bin/modify_vcf_scout.pl#L30-L68

**Comment by alkc** [comment]
Moved the rank model issue to own project here: #150

**Comment by alkc** [comment]
As of this build a standard wgs run does not load into scout without issues:

```
2024-02-06 11:02:21 ab725e716103 scout.parse.variant.transcript[1039] WARNING Failed to parse variant frequencies
2024-02-06 11:02:21 ab725e716103 scout.parse.variant.transcript[1039] WARNING Only splitted and normalised VEP v90+ is supported
2024-02-06 11:02:21 ab725e716103 scout.parse.variant.transcript[1039] WARNING Failed to parse variant frequencies
2024-02-06 11:02:21 ab725e716103 scout.parse.variant.transcript[1039] WARNING Only splitted and normalised VEP v90+ is supported
2024-02-06 11:02:21 ab725e716103 scout.parse.variant.transcript[1039] WARNING Failed to parse variant frequencies
```

**Comment by alkc** [comment]
scout case load for wgs crashes with the following error:

```
2024-02-06 11:07:05 ab725e716103 scout.adapter.mongo.variant_loader[1039] ERROR unexpected error
Traceback (most recent call last):
  File "/home/worker/app/scout/adapter/mongo/variant_loader.py", line 719, in load_variants
    nr_inserted = self._load_variants(
  File "/home/worker/app/scout/adapter/mongo/variant_loader.py", line 436, in _load_variants
    parsed_variant = parse_variant(
  File "/home/worker/app/scout/parse/variant/variant.py", line 177, in parse_variant
    parsed_transcripts = add_gene_and_transcript_info(parsed_variant, variant, vep_header)
  File "/home/worker/app/scout/parse/variant/variant.py", line 561, in add_gene_and_transcript_info
    for parsed_transcript in parse_transcripts(raw_transcripts):
  File "/home/worker/app/scout/parse/variant/transcript.py", line 80, in parse_transcripts
    transcript["region_annotations"] = get_regional_annotation(functional_annotations)
  File "/home/worker/app/scout/parse/variant/transcript.py", line 292, in get_regional_annotation
    regional_list.append(SO_TERMS[annotation]["region"])
KeyError: 'coding_transcript_variant'
2024-02-06 11:07:05 ab725e716103 scout.adapter.mongo.variant_loader[1039] WARNING Deleting inserted variants
```

This is on the demo build of scout on trannel, so slight chance of this issue being specific to this limited build. on the other hand this has never occurred for earlier runs that used vep 103.

**Comment by Jakob37** [comment]
> scout case load for wgs crashes with the following error:
> 
> ```
> 2024-02-06 11:07:05 ab725e716103 scout.adapter.mongo.variant_loader[1039] ERROR unexpected error
> Traceback (most recent call last):
>   File "/home/worker/app/scout/adapter/mongo/variant_loader.py", line 719, in load_variants
>     nr_inserted = self._load_variants(
>   File "/home/worker/app/scout/adapter/mongo/variant_loader.py", line 436, in _load_variants
>     parsed_variant = parse_variant(
>   File "/home/worker/app/scout/parse/variant/variant.py", line 177, in parse_variant
>     parsed_transcripts = add_gene_and_transcript_info(parsed_variant, variant, vep_header)
>   File "/home/worker/app/scout/parse/variant/variant.py", line 561, in add_gene_and_transcript_info
>     for parsed_transcript in parse_transcripts(raw_transcripts):
>   File "/home/worker/app/scout/parse/variant/transcript.py", line 80, in parse_transcripts
>     transcript["region_annotations"] = get_regional_annotation(functional_annotations)
>   File "/home/worker/app/scout/parse/variant/transcript.py", line 292, in get_regional_annotation
>     regional_list.append(SO_TERMS[annotation]["region"])
> KeyError: 'coding_transcript_variant'
> 2024-02-06 11:07:05 ab725e716103 scout.adapter.mongo.variant_loader[1039] WARNING Deleting inserted variants
> ```
> 
> This is on the demo build of scout on trannel, so slight chance of this issue being specific to this limited build. on the other hand this has never occurred for earlier runs that used vep 103.

Might be that Scout is not up to date with the latest VEP? Worth investigating, and reporting to the Scout devs if so

**Comment by alkc** [comment]
> As of this build a standard wgs run does not load into scout without issues.
> 
> So far an endless stream of this:
> 
> ```
> 2024-02-06 11:02:21 ab725e716103 scout.parse.variant.transcript[1039] WARNING Failed to parse variant frequencies
> 2024-02-06 11:02:21 ab725e716103 scout.parse.variant.transcript[1039] WARNING Only splitted and normalised VEP v90+ is supported
> 2024-02-06 11:02:21 ab725e716103 scout.parse.variant.transcript[1039] WARNING Failed to parse variant frequencies
> 2024-02-06 11:02:21 ab725e716103 scout.parse.variant.transcript[1039] WARNING Only splitted and normalised VEP v90+ is supported
> 2024-02-06 11:02:21 ab725e716103 scout.parse.variant.transcript[1039] WARNING Failed to parse variant frequencies
> ```

More data.

```
2024-02-06 12:50:21 918cbc15b667 scout.parse.variant.variant[206] DEBUG iCase ID contains '-'. Using display_name as case _id. 
2024-02-06 12:50:21 918cbc15b667 scout.parse.variant.transcript[206] WARNING Failed to parse variant frequencies               
2024-02-06 12:50:21 918cbc15b667 scout.parse.variant.transcript[206] WARNING Only splitted and normalised VEP v90+ is supported
2024-02-06 12:50:21 918cbc15b667 scout.parse.variant.transcript[206] ERROR could not convert string to float: '0.117696&0.1177'
2024-02-06 12:50:21 918cbc15b667 scout.parse.variant.transcript[206] WARNING Failed to parse variant frequencies               
2024-02-06 12:50:21 918cbc15b667 scout.parse.variant.transcript[206] WARNING Only splitted and normalised VEP v90+ is supported
2024-02-06 12:50:21 918cbc15b667 scout.parse.variant.transcript[206] ERROR could not convert string to float: '0.117696&0.1177'
```

**Comment by alkc** [comment]
> > As of this build a standard wgs run does not load into scout without issues.
> > So far an endless stream of this:
> > ```
> > ...
> > ```
> 
> More data.
> 
> ```
> ...
> 2024-02-06 12:50:21 918cbc15b667 scout.parse.variant.transcript[206] ERROR could not convert string to float: '0.117696&0.1177'
> ...
> ```

Looks like something's up with the new Gnomad where the offending values lie:

## VEP 111

```
gnomADe_AF                    0.117696&0.1177
gnomADe_AFR_AF                0.1006
gnomADe_AMR_AF                0.08657
gnomADe_ASJ_AF                0.05427
gnomADe_EAS_AF                0.3203
gnomADe_FIN_AF                0.04098
gnomADe_NFE_AF                0.08343
gnomADe_OTH_AF                0.09743
gnomADe_SAS_AF                0.2468
gnomADg_AF                    0.0905672&0.09823
gnomADg_AFR_AF                0.09848
gnomADg_AMI_AF                0.2292
gnomADg_AMR_AF                0.07573
gnomADg_ASJ_AF                0.05616
gnomADg_EAS_AF                0.3107
gnomADg_FIN_AF                0.04246
gnomADg_MID_AF                0.1013
gnomADg_NFE_AF                0.08469
gnomADg_OTH_AF                0.1076
gnomADg_SAS_AF                0.2543
MAX_AF                        0.3203
```

## VEP 103

```
gnomAD_AF                     0.1177
gnomAD_AFR_AF                 0.1006
gnomAD_AMR_AF                 0.08657
gnomAD_ASJ_AF                 0.05427
gnomAD_EAS_AF                 0.3203
gnomAD_FIN_AF                 0.04098
gnomAD_NFE_AF                 0.08343
gnomAD_OTH_AF                 0.09743
gnomAD_SAS_AF                 0.2468
MAX_AF                        0.3203
MAX_AF_POPS                   gnomAD_EAS
```

Looking into this.

**Comment by alkc** [comment]
Commit [85b8a9](https://github.com/Clinical-Genomics-Lund/nextflow_wgs/pull/132/commits/85b8a9bf0804e78f433577d7efa6ba75a09d4259) eliminates both errors when loading into scout.

Moving onto onco samples.

**Comment by alkc** [comment]
Retrying WGS runs w/ updated `modify_scout_vcf.pl` internal ranking. 

Some of the SO terms that have started to bubble up need to be added to scout.  Ongoing at: https://github.com/Clinical-Genomics/scout/pull/4434

In the meantime I'll try loading a WGS trio/single into trannel-scout running on the dev branch associated w/ that PR linked above.

**Comment by alkc** [comment]
rerunning samples after patching a new bug in modify scout script.

`oncov2` finishes and loads into scout. small variations in rank score on first glance.

will retry w/ wgs trio and singles before running final checks.

File: configs/nextflow.hopper.config [hunk 1]
```diff
@@ -44,17 +43,22 @@ params {
   rCRS_fasta        = "${refpath}/fasta/split/M.fa"
 
   // VEP REFERENCES AND ANNOTATION DBS //
+  VEP_DIR           = "${refpath}/vep/.vep111.0"
```
Comment (Jakob37) [inline configs/nextflow.hopper.config:46]: I think we should consider renaming the dir to `vep111.0`. It has confused me many times while looking through the annotation files where the vep files are

File: configs/nextflow.hopper.config [hunk 2]
```diff
@@ -44,17 +43,22 @@ params {
...
+  VEP_CACHE         = "${VEP_DIR}"
+  VEP_PLUGINS       = "${VEP_DIR}/Plugins"
+  VEP_SYNONYMS      = "${VEP_DIR}/synonyms"
+  VEP_FASTA         = "${VEP_DIR}/homo_sapiens_merged/111_GRCh38/Homo_sapiens.GRCh38.dna.toplevel.fa.gz"
+  VEP_MAX_SV_SIZE   = "50000000"
```
Comment (Jakob37) [inline configs/nextflow.hopper.config:51]: Are these two the default values?

File: configs/nextflow.hopper.config [hunk 3]
```diff
@@ -44,17 +43,22 @@ params {
...
+  VEP_DISTANCE      = "200"
```
Comment (Jakob37) [inline configs/nextflow.hopper.config:52]: Rename to "VEP_TRANSCRIPT_DISTANCE" to make more self explanatory?

**Review by Jakob37 — COMMENTED** [review]
Nice!

**Review by Jakob37 — APPROVED** [review]
I think it looks OK to proceed to the validation when your testing of this is done. So I'll approve.

**Comment by alkc** [comment]
Thanks. I've one last test run that died together with the cluster. I'll merge once that one's done and the rank scores look OK

File: configs/nextflow.hopper.config (hunk 2 above)
Comment (alkc) [reply configs/nextflow.hopper.config:51]: nyet. MAX_SV_SIZE is 5k. distance doesn't say.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,17 @@
 # CHANGELOG
 
+## [Unreleased]
```
Comment (Jakob37) [inline CHANGELOG.md:3]: Nice with the cleanups in the changelog!

If you want to keep it really tidy, you could keep consistent line endings as well (i e not mixing ending with period and not ending with period)

File: bin/cleanVCF.py [hunk 5]
```diff
@@ -1,16 +1,59 @@
...
+with open(args.vcf) as vcf:
+    for idx, line in enumerate(vcf):
+        line = line.strip()
+
+        if line[0] == "#":
+            print(line)
+            continue
+
+        content = line.split("\t")
+        curr_var = " ".join(content[0:2] + content[5:7])
```
Comment (Jakob37) [inline bin/cleanVCF.py:45]: I would prefer naming these ranges for readability

For instance `content[6]` vs 

```
info_col_i = 6
content[info_col_i]
```

Comment (Jakob37) [reply bin/cleanVCF.py:45]: Sorry, `filter_col_i` in this case I guess!

**Review by Jakob37 — APPROVED** [review]
Great job, and endurance on this!

We learned a lot on the way.

Looking forward to start finalizing the validation.

File: bin/cleanVCF.py (hunk 5 above)
Comment (alkc) [reply bin/cleanVCF.py:48]: Fixed!

**Pull request merged**